shell/platform/darwin/ios/ios_surface_metal.mm

```diff
@@ -3,110 +3,65 @@
 // found in the LICENSE file.
 
 #include "flutter/shell/platform/darwin/ios/ios_surface_metal.h"
+
 #include "flutter/shell/gpu/gpu_surface_metal.h"
+#include "flutter/shell/platform/darwin/ios/ios_context_metal.h"
 
 namespace flutter {
 
-IOSSurfaceMetal::IOSSurfaceMetal(fml::scoped_nsobject<CAMetalLayer> layer,
-                                 FlutterPlatformViewsController* platform_views_controller)
-    : IOSSurface(platform_views_controller), layer_(std::move(layer)) {}
-
-IOSSurfaceMetal::IOSSurfaceMetal(fml::scoped_nsobject<CAMetalLayer> layer)
-    : IOSSurface(nullptr), layer_(std::move(layer)) {}
-
-IOSSurfaceMetal::~IOSSurfaceMetal() = default;
-
-// |IOSSurface|
-bool IOSSurfaceMetal::IsValid() const {
-  return layer_;
+static fml::scoped_nsobject<CAMetalLayer> CastToMetalLayer(fml::scoped_nsobject<CALayer> layer) {
+  FML_DCHECK([layer isKindOfClass:[CAMetalLayer class]]);
+  return fml::scoped_nsobject<CAMetalLayer>{reinterpret_cast<CAMetalLayer*>([layer.get() retain])};
 }
 
-// |IOSSurface|
-bool IOSSurfaceMetal::ResourceContextMakeCurrent() {
-  return false;
+static IOSContextMetal* CastToMetalContext(const std::shared_ptr<IOSContext>& context) {
+  return reinterpret_cast<IOSContextMetal*>(context.get());
 }
 
-// |IOSSurface|
-void IOSSurfaceMetal::UpdateStorageSizeIfNecessary() {}
-
-// |IOSSurface|
-std::unique_ptr<Surface> IOSSurfaceMetal::CreateGPUSurface(GrContext* gr_context) {
-  if (gr_context) {
-    return std::make_unique<GPUSurfaceMetal>(this, sk_ref_sp(gr_context), layer_);
+IOSSurfaceMetal::IOSSurfaceMetal(fml::scoped_nsobject<CALayer> layer,
+                                 std::shared_ptr<IOSContext> context,
+                                 FlutterPlatformViewsController* platform_views_controller)
+    : IOSSurface(std::move(context), platform_views_controller),
+      layer_(CastToMetalLayer(std::move(layer))) {
```

Comment:
same note about casting

Reply:
Now all the casts are in the IOSSurface<Platform>. Do you want them higher?

Reply:
Here is my thought: It's silly to cast something in a constructor because someone should have and know the requirements to make an object.  This just hides the cast better and casts are always something that should be regarded with skepticism since they can be the source of nasty bugs.

I'd say this is something similar to [demeter's law](https://en.wikipedia.org/wiki/Law_of_Demeter).  where you would prefer `int foo(Child x) => x.foo` instead of `int foo(Parent x) => x.child.foo`.